### PR TITLE
Add latch header implementation

### LINKED ISSUE #67: C++20 feature: latch
Issue tracking implementation of `latch`es, introduced in C++20.

See [latch](https://en.cppreference.com/w/cpp/thread/latch) on cppreference.com.

Potential implementations:
- Atomic-based. Portable, but does not cause threads to sleep when blocking.
- Atomic variable & `condition_variable`. Mostly efficient. May have issues with cost of construction / destruction.
- Atomic & [WaitOnAddress](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress). Requires Windows 8, but should have minimal overhead.

### PR DESCRIPTION
**Pull request by diamante0018**
Hello, this is my first contribution to this repository. I saw there are open issues for some C++20 features and since `latch` looked pretty easy I gave it a go. As suggested by the person that created the issue I used `<atomic>` based implementation. This means it may render this feature unusable on Windows versions older than Vista. However, it greatly simplifies the code and it seems it is the choice of other C++ STL implementations.

I tried my best to follow the conventions of this repository but I invite any reviewer to check if I am missing any special header guard, especially at the bottom of the file to check in an implementation is already available.

- [x] #67
- [x] Add basic test coverage
- [x] Add the header to the various tools used to generate `<latch>` (to double-check if everything is included)
- [x] Reviewer needs to double-check for missing special MINGW header guards in the `<latch>` header

Closes #67

### CODE REVIEW

File: mingw.latch.h
Comment (nmcclatchey) [inline mingw.latch.h:82]: This is not a sufficient condition. The call to `atomic::wait()` may unblock with a nonzero counter in the following circumstances:
* Underlying value has changed (at least one call to an unblocking function has been performed), and
* The implementation (of the atomic) has unblocked spuriously.
The [specification](https://en.cppreference.com/w/cpp/atomic/atomic/wait) for `std::atomic`'s `wait()` call is only guaranteed not to return spuriously if the value of the atomic has kept the same value (that is, it does not guarantee any specific value).

Additionally, there is a race condition regardless of whether the atomic implementation can unblock spuriously, because the value in the atomic counter could have changed to a nonzero value between the check of its value and the call to wait. In this implementation, that would result in `latch::wait()` returning despite not all awaited tasks being completed.

To resolve these, the `latch::wait()` function should test and wait on its atomic counter in a loop, only exiting when the counter has been verified to have reached a non-positive value.

File: mingw.latch.h
Comment (nmcclatchey) [inline mingw.latch.h:77]: Testing whether the counter is less than or equal to 0 may improve behavior in the case of a user having bugs in their code.

File: mingw.latch.h
Comment (nmcclatchey) [inline mingw.latch.h:90]: As with `count_down()`, there is a bug to resolve here, due to `atomic::fetch_sub()` returning the value from before the subtraction. The same fix, testing `current == update` instead of `current == 0`, will also fix it.

Likewise, the same recommended improvement, testing `current <= update`, will make the library more forgiving to users with bugs.

File: mingw.latch.h
Comment (nmcclatchey) [inline mingw.latch.h:98]: Because `fetch_sub()` returns the value from before the subtraction, the value on which you want to wait is actually `current - update`.

File: mingw.latch.h
Comment (nmcclatchey) [inline mingw.latch.h:90]: As with `count_down()`, I recommend asserting that `update <= current`.

File: tests/tests.cpp
Comment (nmcclatchey) [inline tests/tests.cpp:313]: Perhaps also add a test for `try_wait()`?

Comment (nmcclatchey) [inline tests/tests.cpp:313]: You can test for `noexcept` specifiers using `static_assert(noexcept(expression), "message")`. In the case of `latch`, `max()` and `try_wait()` are both `noexcept`.

File: tests/tests.cpp
Comment (nmcclatchey) [inline tests/tests.cpp:7]: Inclusion of and tests for `latch` should be gated behind the necessary `__cplusplus` version. In particular, the implemented `latch` requires a C++20 compiler.

Without this conditional inclusion and compilation, the test suite will no longer function on older compilers or standard versions.

File: mingw.latch.h
Comment (nmcclatchey) [inline mingw.latch.h:31]: This implementation of `latch` does not directly use any features from Windows. As such, it should not be gated behind a Windows version.

**Review by nmcclatchey — CHANGES_REQUESTED** [review]
Thank you for implementing this. It will be a welcome addition to the mingw-std-threads library. However, this implementation has two bugs that should be resolved before merging, and will unintentionally break tests on older compilers with its update to the test suite. Additionally, there are some opportunities to add (optional) quality-of-life improvements for users (specifically, alerting them when certain bugs occur in debug mode, and behaving a bit more like one might expect if bugs are present in release mode).

**Comment by nmcclatchey** [comment]
@diamante0018 , please also note that while I have been a frequent contributor, I am not a maintainer; specifically, I am unable to merge pull requests. When the bugs mentioned in my review have been resolved, I recommend alerting [alxvasilev](https://github.com/alxvasilev).

**Comment by alxvasilev** [comment]
Thanks for reviewing this, @nmcclatchey!

**Comment by diamante0018** [comment]
Thanks for the in-depth review. I will try to address it tomorrow.

File: mingw.latch.h
Comment (diamante0018) [reply mingw.latch.h:63]: So I should check for `current == update` and also remove `- update`

**Comment by diamante0018** [comment]
Hello @alxvasilev, I should have addressed all requested changes. I did not squash the commits down as I was told when contributing to some other project that it makes reviewing harder. I will leave it up to you whenever squashing before merging the PR is the best move.
But before that, please make sure I did not miss anything. With the newly added tests and bugs removed, this should be good to go. Thanks a lot again @nmcclatchey

File: mingw.latch.h
Comment (nmcclatchey) [reply mingw.latch.h]: I had completely missed the `- update`. The bug was in my analysis, rather than your code.

File: mingw.latch.h
Comment (nmcclatchey) [inline mingw.latch.h:82]: Typo here; this should be `<=`.

**Review by nmcclatchey — CHANGES_REQUESTED** [review]
One typo remains.

File: mingw.latch.h
Comment (diamante0018) [reply mingw.latch.h]: Understood. Is it fine now or should I revert?

File: mingw.latch.h
Comment (diamante0018) [reply mingw.latch.h]: Done

File: mingw.latch.h
Comment (nmcclatchey) [reply mingw.latch.h]: It is fine now. Now that I have seen the `- update` in your original code, I can safely say that which variant you select is a matter of taste, rather than of correctness.

**Comment by diamante0018** [comment]
@alxvasilev Looks like this PR is ready for your review too. Thanks 🐱

**Comment by diamante0018** [comment]
@alxvasilev bump

**Comment by diamante0018** [comment]
Thank you very much!

**Comment by alxvasilev** [comment]
Thanks @diamante0018 for your contribution, it's really appreciated!
As always, big thanks to @nmcclatchey for your dedication to this project and doing most of the work!

**Pull request merged**
